File: Part1_강의자료/Ch08_투포인터/문제별코드/1806_부분합/1806_binary_search.py
N, S = map(int, input().split())
li = [0] + list(map(int, input().split()))
acc = [0] * (N + 1)
for i in range(1, N + 1):
    acc[i] = acc[i - 1] + li[i]

ans = N + 1
for i in range(1, N + 1):
    l, r = i, N
    while l <= r:
        m = (l + r) // 2
        current_sum = acc[m] - acc[i - 1]
        if current_sum >= S:
            ans = min(ans, m - i + 1)
            r = m - 1
        else:
            l = m + 1

print(0 if ans == N + 1 else ans)

File: Part1_강의자료/Ch08_투포인터/문제별코드/1806_부분합/1806_binary_search.cpp
#include <algorithm>
#include <stdio.h>
#include <vector>
using namespace std;

int main()
{
    int N, S;
    long long num;
    scanf("%d %d", &N, &S);

    vector<long long> acc(N + 1);
    for (int i = 1; i <= N; i++)
    {
        scanf("%lld", &num);
        acc[i] = acc[i - 1] + num;
    }

    int ans = N + 1;
    for (int i = 1; i <= N; i++)
    {
        int l = i, r = N;
        while (l <= r)
        {
            int m = (l + r) / 2;
            long long sum = acc[m] - acc[i - 1];

            if (sum >= S)
            {
                ans = min(ans, m - i + 1);
                r = m - 1;
            }
            else
                l = m + 1;
        }
    }

    printf("%d\n", ans == N + 1 ? 0 : ans);
    return 0;
}
